pkg/parse/state.go

```diff
@@ -45,46 +52,102 @@ type ReconcilerState struct {
 	syncErrorCache *SyncErrorCache
 }
 
-func (s *ReconcilerState) checkpoint() {
-	applied := s.cache.source.syncPath.OSPath()
-	if applied == s.checkpointedSyncPath {
-		return
+type checkpoint struct {
+	// sourcePath caches the source path that was last successfully applied.
+	// Set to the empty string if the last attempt failed.
+	sourcePath cmpath.Absolute
+	// lastUpdateTime is the last time the checkpoint was updated.
+	// AKA: Last successful sync.
+	// TODO: Surface this timestamp in the RSync status API
+	lastUpdateTime metav1.Time
+	// lastTransitionTime is the last time the checkpoint was updated with a new sourcePath.
+	// AKA: First successful sync with this sourcePath (repo + branch + source commit + syncDir).
+	// TODO: Surface this timestamp in the RSync status API
+	lastTransitionTime metav1.Time
+}
+
+// updateCheckpoint records the last known source path, updates the
+// timestamps, and logs the message.
+func (s *ReconcilerState) updateCheckpoint(c clock.Clock, newSourcePath cmpath.Absolute) {
+	now := nowMeta(c)
+	// Check for transition
+	transitioned := false
+	if s.checkpoint.sourcePath != newSourcePath {
+		transitioned = true
+		s.checkpoint.sourcePath = newSourcePath
+		s.checkpoint.lastTransitionTime = now
 	}
-	klog.Infof("Reconciler checkpoint updated to %s", applied)
-	s.checkpointedSyncPath = applied
+	// Record when the checkpoint was last updated
+	s.checkpoint.lastUpdateTime = now
+	if newSourcePath == invalidSyncPath {
+		klog.Info("Reconciler checkpoint invalidated")
+	} else if transitioned {
+		klog.Infof("Reconciler checkpoint updated with new source path: %s", newSourcePath)
+	} else {
+		klog.Infof("Reconciler checkpoint updated with existing source path: %s", newSourcePath)
+	}
+}
+
+// RecordSyncSuccess is called after a successful sync. It records the last
+// known source path, as well as timestamps for last updated and last
+// transitioned.
+func (s *ReconcilerState) RecordSyncSuccess(c clock.Clock) {
+	klog.Info("Sync successful")
+	s.updateCheckpoint(c, s.cache.source.syncPath)
 	s.cache.needToRetry = false
 }
 
-// invalidate logs the errors, clears the state tracking information.
-// invalidate does not clean up the `s.cache`.
-func (s *ReconcilerState) invalidate(errs status.MultiError) {
+// RecordRenderInProgress is called when waiting for rendering status. It
+// invalidates the checkpoint, requests a retry, and resets the cacheForCommit,
+// which tells the retry to re-parse from source.
+func (s *ReconcilerState) RecordRenderInProgress(c clock.Clock) {
+	klog.Info("Rendering in progress")
+	s.updateCheckpoint(c, invalidSyncPath)
```

Comment:
The function only needs to reset cache `s.cache = cacheForCommit`.
The caller returns an error, which will call `RecordFailure` to updateCheckpoint and set s.cache.needToRetry to true.

Reply:
Good catch. I swear I had it doing that at some point. Musta lost it in one of the copies to a new PR.

Reply:
fixed